File: android/src/main/java/com/reactnative/unity/view/UnityMessageEvent.java
package com.reactnative.unity.view;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.events.Event;
import com.facebook.react.uimanager.events.RCTEventEmitter;

/**
 * Created by xzper on 2018-03-08.
 */

public class UnityMessageEvent extends Event<UnityMessageEvent> {

    public static final String EVENT_NAME = "unityMessage";
    private final String mData;

    public UnityMessageEvent(int viewId, String data) {
        super(viewId);
        mData = data;
    }

    @Override
    public String getEventName() {
        return EVENT_NAME;
    }

    @Override
    public void dispatch(RCTEventEmitter rctEventEmitter) {
        WritableMap data = Arguments.createMap();
        data.putString("message", mData);
        rctEventEmitter.receiveEvent(getViewTag(), EVENT_NAME, data);
    }
}

File: windows/RNUnityView/UnityMessageEvent.cs
using Newtonsoft.Json.Linq;
using ReactNative.UIManager.Events;

namespace RNUnityView
{
    public sealed class UnityMessageEvent : Event
    {
        public const string EVENT_NAME = "unityMessage";

        private readonly string data;

        public UnityMessageEvent(int viewId, string data)
            : base(viewId)
        {
            this.data = data;
        }

        public override string EventName => EVENT_NAME;

        public override void Dispatch(RCTEventEmitter eventEmitter)
        {
            var eventData = new JObject
            {
                { "message", data },
                { "target", ViewTag },
            };

            eventEmitter.receiveEvent(ViewTag, EventName, eventData);
        }
    }
}
